### PR TITLE
fix(context): reliably fire BrowserContext.Close event when browser is closing

### PR DESCRIPTION
**Pull request by yury-s**

### CODE REVIEW

File: src/server/pipeTransport.ts [hunk 1]
```diff
@@ -24,6 +24,9 @@ export class PipeTransport implements ConnectionTransport {
   private _pendingMessage = '';
   private _eventListeners: RegisteredListener[];
   private _waitForNextTask = makeWaitForNextTask();
+
+  // This method must be overridden to provide custom close behavior.
+  close: () => void = () => { throw new Error('Pipe transport cannot be closed by default'); };
```
Comment (dgozman) [inline src/server/pipeTransport.ts]: Let's pass closeHandler to constructor?

File: src/server/pipeTransport.ts [hunk 2]
```diff
@@ -72,11 +75,4 @@ export class PipeTransport implements ConnectionTransport {
     }
     this._pendingMessage = buffer.toString(undefined, start);
   }
-
-  close() {
-    this._pipeWrite = null;
-    helper.removeEventListeners(this._eventListeners);
```
Comment (dgozman) [inline src/server/pipeTransport.ts]: Let's remove listeners in line 28.

**Review by dgozman — APPROVED** [review]

File: src/server/pipeTransport.ts (hunk 1 above)
Comment (yury-s) [reply src/server/pipeTransport.ts]: Done.

File: src/server/pipeTransport.ts (hunk 2 above)
Comment (yury-s) [reply src/server/pipeTransport.ts]: Done.

File: src/server/pipeTransport.ts [hunk 3]
```diff
@@ -24,14 +24,19 @@ export class PipeTransport implements ConnectionTransport {
   private _pendingMessage = '';
   private _eventListeners: RegisteredListener[];
   private _waitForNextTask = makeWaitForNextTask();
+
+  // This method must be overridden to provide custom close behavior.
```
Comment (dgozman) [inline src/server/pipeTransport.ts]: This comment is outdated now.

File: src/server/pipeTransport.ts [hunk 4]
```diff
@@ -72,11 +77,4 @@ export class PipeTransport implements ConnectionTransport {
     }
     this._pendingMessage = buffer.toString(undefined, start);
   }
-
```
Comment (dgozman) [inline src/server/pipeTransport.ts]: Why don't we keep close?

```
close() {
  this._closeCallback();
}
```

File: src/server/webkit.ts [hunk 5]
```diff
@@ -378,7 +377,7 @@ function wrapTransportWithWebSocket(transport: ConnectionTransport, port: number
         pendingBrowserContextDeletions.set(seqNum, params.browserContextId);
     });
 
-    socket.on('close', () => {
+    socket.on('close', closeListener = () => {
```
Comment (dgozman) [inline src/server/webkit.ts]: socket.__closeListener = ....;

**Review by dgozman — APPROVED** [review]

File: src/server/pipeTransport.ts (hunk 3 above)
Comment (yury-s) [reply src/server/pipeTransport.ts]: Removed

File: src/server/pipeTransport.ts (hunk 4 above)
Comment (yury-s) [reply src/server/pipeTransport.ts]: Done.

File: src/server/webkit.ts (hunk 5 above)
Comment (yury-s) [reply src/server/webkit.ts]: Done.

**Pull request merged**